### PR TITLE
Customization for _getUploadRoot and _getAllowedExtensions in Mage_Adminhtml_Model_System_Config_Backend_File

### PR DESCRIPTION
**Pull request by eneiasramos**
…lowedExtensions

### Description (*)

Here we have the implementation of 2 functionalities:

1 - Add validation for allowed extensions when uploading files.

today:

```
<upload_dir config="system/filesystem/media" scope_info="1">sales/store/logo</upload_dir>
```

new:

```
<upload_dir allowed_extensions="png,jpg,jpeg,gif" config="system/filesystem/media" scope_info="1">sales/store/logo</upload_dir>
```

2 - Add validation to the config's attribute in upload_dir's tag (not used today). Also the possibility of customizing it. Example:

today:
```
<upload_dir config="system/filesystem/media" scope_info="1">docs/pdf</upload_dir>
```

new:

```
<upload_dir allowed_extensions="pdf" config="system/filesystem/var" scope_info="1">docs/pdf</upload_dir>
```

### Fixed Issues (if relevant)

1. Fixes OpenMage/magento-lts#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All automated tests passed successfully (all builds are green)
 - [ ] Add yourself to contributors list

### CODE REVIEW

**Comment by fballiano** [comment]
@eneiasramos could you please write a real title and description? thanks!

**Comment by eneiasramos** [comment]
> @eneiasramos could you please write a real title and description? thanks!

Sorry I forgot that my dear :smile_cat:

**Comment by eneiasramos** [comment]
@fballiano done my dear!

**Comment by fballiano** [comment]
@Flyingmana can you help me here? I'm not sure I understand how to test this

**Comment by Flyingmana** [comment]
will have a look on the weekend

**Comment by pquerner** [comment]
Images can use the `backend_model` `Mage_Adminhtml_Model_System_Config_Backend_Image`, this will allow uploading of  `'jpg', 'jpeg', 'gif', 'png'` already. No need to have that in the base class. 

As for the path I don't follow the changes here. Can you explain some more of what you are trying to do?

If you only want to allow a specific file extension I would provide a class that extends `Mage_Adminhtml_Model_System_Config_Backend_File` and override `protected function _getAllowedExtensions`.
(see `Mage_Adminhtml_Model_System_Config_Backend_Image` as an example)

**Comment by Flyingmana** [comment]
Ok, for the how to Test part.

have a config which overwrites the system.xml for 
image

There by adding the new `allowed_extensions` attribute it should be able to control which ones are allowed to upload.

Although, its probably more for the logo in app/code/core/Mage/Sales/etc/system.xml not the placeholder one in my screenshot.


What Iam not sure about is, what the exact use case or benefit should be.

Also as this part is a bit complex, I would would like to require a phpunit test for this parts

File: app/code/core/Mage/Adminhtml/Model/System/Config/Backend/File.php [hunk 1]
```diff
@@ -144,6 +146,12 @@ protected function _getUploadDir()
      */
     protected function _getUploadRoot($token)
     {
+        $value = Mage::getStoreConfig($token);
+        if (preg_match(self::SYSTEM_FILESYSTEM_REGEX, $value, $matches) !== false) {
+            $dir = str_replace('root_dir', 'base_dir', $matches[1]);
+            $path = str_replace('/', DS, $matches[2]);
+            return Mage::getConfig()->getOptions()->getData($dir) . $path;
+        }
```
Comment (Flyingmana) [inline app/code/core/Mage/Adminhtml/Model/System/Config/Backend/File.php:154]: please revert the changes in this method(and do it as a separate PullRequest with proper description), they are not covered by the Description and there is no reason visible for me, why this change might be needed.

**Review by Flyingmana — CHANGES_REQUESTED** [review]
please do separate the getUploadRoot change, and provide further explaination why the getAllowedExtensions should be extended in this way and which file formats you actually plan to add (or remove?)

**Comment by eneiasramos** [comment]
@Flyingmana here:

https://github.com/OpenMage/magento-lts/pull/4078

https://github.com/OpenMage/magento-lts/pull/4079

File: app/code/core/Mage/Adminhtml/Model/System/Config/Backend/File.php (hunk 1 above)
Comment (eneiasramos) [reply app/code/core/Mage/Adminhtml/Model/System/Config/Backend/File.php:154]: I put it in the comment below

**Comment by fballiano** [comment]
replaced by https://github.com/OpenMage/magento-lts/pull/4079 and https://github.com/OpenMage/magento-lts/pull/4078/files
